Add quirk to allow downloads on dropbox.com

achristensen07 opened the pull request:
#### df7d7806ba4886c1e40f251f4566661d87165593
<pre>
Add quirk to allow downloads on dropbox.com
<a href="https://bugs.webkit.org/show_bug.cgi?id=296749">https://bugs.webkit.org/show_bug.cgi?id=296749</a>
<a href="https://rdar.apple.com/155649992">rdar://155649992</a>

Reviewed by Brent Fulgham.

<a href="https://www.dropbox.com/download?os=mac">https://www.dropbox.com/download?os=mac</a> has a content-security-policy sandbox that does not
allow downloads, but only if requested with a Safari user-agent.  This was not an issue until
we increased our standards compliance and interoperability and security in 283667@main by
honoring content-security-policy sandboxes with regards to downloads initiated by the main frame.
Unfortunately, this broke the ability to download the dropbox installer, which dropbox.com is
saying should not be allowed on its website.  Add a quirk to get the old behavior until they fix it.

* Source/WebCore/loader/DocumentLoader.cpp:
(WebCore::DocumentLoader::continueAfterContentPolicy):
* Source/WebCore/page/Quirks.cpp:
(WebCore::Quirks::shouldAllowDownloadsInSpiteOfCSP const):
* Source/WebCore/page/Quirks.h:

Canonical link: <a href="https://commits.webkit.org/298093@main">https://commits.webkit.org/298093@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6306e41e2881b1bb52c788291cd344d63fafa529

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/114224 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/33971 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/24433 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/120390 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/64960 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/116113 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/34600 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/42533 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/86807 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/64960 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/117172 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/27554 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/102587 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/67196 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/26736 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/20713 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/64085 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/96928 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/20830 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/123607 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/41243 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/30751 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/95634 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/41620 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/98789 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/95417 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/40559 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/18386 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/37329 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/18305 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/41122 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/46633 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/40729 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/44035 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/42480 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

brentfulgham commented on Source/WebCore/page/Quirks.cpp:
```diff
@@ -1542,6 +1542,15 @@ bool Quirks::shouldUseEphemeralPartitionedStorageForDOMCookies(const URL& url) c
+    if (!needsQuirks())
+        return false;
+
+    return isDomain("dropbox.com"_s);
```
It would be more efficient if you added a new flag to `QuirksData`, that is set in `Quirks::determineRelevantQuirks()` so that we do not have to do this string comparison every time we want to check if this Quirk is needed in the DOM.

achristensen07 replied:
I verified that actually doesn't work because at this point in the loading process Quirks::determineRelevantQuirks just returns early.  As it is, this is only one comparison per navigation that we decide to download.

brentfulgham replied:
Great! Then ship it!

brentfulgham reviewed (approved):
I'd prefer if you handled the Quirk check more efficiently (as I shared inline), but I can do a follow-up cleanup if you need to get this landed ASAP.

achristensen07 committed:
Add quirk to allow downloads on dropbox.com
https://bugs.webkit.org/show_bug.cgi?id=296749
rdar://155649992

Reviewed by Brent Fulgham.

https://www.dropbox.com/download?os=mac has a content-security-policy sandbox that does not
allow downloads, but only if requested with a Safari user-agent.  This was not an issue until
we increased our standards compliance and interoperability and security in 283667@main by
honoring content-security-policy sandboxes with regards to downloads initiated by the main frame.
Unfortunately, this broke the ability to download the dropbox installer, which dropbox.com is
saying should not be allowed on its website.  Add a quirk to get the old behavior until they fix it.

* Source/WebCore/loader/DocumentLoader.cpp:
(WebCore::DocumentLoader::continueAfterContentPolicy):
* Source/WebCore/page/Quirks.cpp:
(WebCore::Quirks::shouldAllowDownloadsInSpiteOfCSP const):
* Source/WebCore/page/Quirks.h:

Canonical link: https://commits.webkit.org/298093@main
